timer/Timer.py

```diff
@@ -22,48 +23,54 @@ class Timer:
     the timer notification is closed.
     """
 
-    def __init__(self, run_seconds, name, callback, persistent=False):
+    def __init__(self, run_seconds, name, callback, loop, persistent=False):
         self.id = next(ids)
         self.run_seconds = run_seconds
         self.name = name
         self.callback = callback
+        self.loop = loop
         self.persistent = persistent
         self.tag = None
         self.end_time = datetime.now() + timedelta(seconds=run_seconds)
         self.notification = None
         self.intervals = [30, 30, 30, 10, 10, 10]
 
-    def start(self, loop):
+    def start(self):
         def on_end(on_close=None):
             self.notify(self.name, self.time_since_end, on_close=on_close)
             if self.persistent:
                 interval = self.intervals.pop() if self.intervals else 60
-                self.tag = loop.call_after_delay(interval, on_end)
+                log.debug(f"Notification sleeping for {interval} seconds...")
+                self.tag = self.loop.call_after_delay(interval, on_end)
             else:
                 self.callback(self)
                 self.tag = None
 
-        def on_close(arg):
-            log.debug("notification closed")
-            self.callback(self)
-            self.stop(loop)
-
-        self.tag = loop.call_after_delay(self.run_seconds, on_end, on_close)
+        self.tag = self.loop.call_after_delay(self.run_seconds, on_end, self.on_notification_close)
         self.notify(self.description, sound=False)
         log.debug('Timer set for %s', self.end_time)
 
-    def stop(self, loop, notify=False):
+    def on_notification_close(self, arg):
+        # https://gnome.pages.gitlab.gnome.org/libnotify/enum.ClosedReason.html
+        reason = self.notification.get_closed_reason()  # 1: timeout, 2: dismissed by the user
+        log.debug("notification closed" + (" by user" if reason == 2 else ''))
+        if reason == 2:
+            # Closed by user, stopping this timer
+            self.stop()
+            self.callback(self)
+
+    def stop(self, notify=False):
         if self.tag is not None:
             self.persistent = False  # prevent unstoppable on_end() calls
-            loop.cancel_callback(self.tag)
+            self.loop.cancel_callback(self.tag)
             if notify:
                 self.notify("Timer stopped", self.description, sound=False)
-                loop.call_after_delay(5, self.notification.close)
+                self.loop.call_after_delay(5, self.notification.close)
             self.tag = None
 
     @property
     def description(self):
-        return f"{self.name} at {self.end_time.strftime('%-I:%M %p')}"
+        return f"{self.name} at {self.end_time.strftime('%-H:%M %p')}"
```

Comment:
@brunetton does this break the AM/PM format?
Some people may be using it, I assume